File: apps/runtime/src/collab/transcript.rs
//! Locating and rendering another agent's transcript — plan §5.6.
//!
//! Every CLI writes its history somewhere different, and none of them promise
//! the shape. So this module is deliberately forgiving: it reads the tail of a
//! file, renders the lines it recognises, and skips the ones it does not. A
//! transcript we can only half read is far more useful to the agent asking than
//! an error, and a transcript we cannot find at all is reported in one sentence
//! rather than as a failure.

use std::{
    fs::File,
    io::{Read, Seek, SeekFrom},
    path::{Path, PathBuf},
};

use serde_json::Value;

/// Only the tail of a transcript is read; a long session is megabytes of JSON.
pub const MAX_TAIL_BYTES: u64 = 5 * 1024 * 1024;
/// `transcript` renders everything it found, up to this many bytes of prose.
pub const MAX_RENDERED_BYTES: usize = 200 * 1024;
/// A tool's input is quoted, not dumped.
const MAX_TOOL_DETAIL: usize = 120;
/// One rendered message line is trimmed to this before it reaches the agent.
const MAX_LINE: usize = 2_000;
/// Ceiling on how many directory entries a session-id search will look at.
const MAX_SCAN_ENTRIES: usize = 20_000;
const MAX_SCAN_DEPTH: usize = 6;

/// Where a transcript came from, so the reply can say so.
#[derive(Debug, Clone)]
pub struct Located {
    pub path: PathBuf,
    /// Human sentence naming the provider and file.
    pub origin: String,
}

/// Reads at most the last [`MAX_TAIL_BYTES`] of a file, starting at the first
/// newline inside the window so the first line is never a fragment.
pub fn read_tail(path: &Path, max_bytes: u64) -> std::io::Result<String> {
    let mut file = File::open(path)?;
    let length = file.metadata()?.len();
    let start = length.saturating_sub(max_bytes);
    if start > 0 {
        file.seek(SeekFrom::Start(start))?;
    }
    let mut buffer = Vec::with_capacity(length.saturating_sub(start).min(max_bytes) as usize);
    file.take(max_bytes).read_to_end(&mut buffer)?;
    let text = String::from_utf8_lossy(&buffer).into_owned();
    if start == 0 {
        return Ok(text);
    }
    Ok(match text.find('\n') {
        Some(index) => text[index + 1..].to_owned(),
        None => String::new(),
    })
}

/// Renders JSONL (or a JSON array / object of messages) into one line per
/// message. Unknown lines are skipped rather than reported.
pub fn render(text: &str) -> Vec<String> {
    // A whole-file JSON document (gemini writes one) is unwrapped first.
    let trimmed = text.trim_start();
    if (trimmed.starts_with('[') || trimmed.starts_with('{'))
        && let Ok(value) = serde_json::from_str::<Value>(trimmed)
    {
        let rendered = render_document(&value);
        if !rendered.is_empty() {
            return rendered;
        }
    }
    text.lines()
        .filter_map(|line| {
            let line = line.trim();
            if line.is_empty() {
                return None;
            }
            serde_json::from_str::<Value>(line).ok()
        })
        .filter_map(|value| render_entry(&value))
        .collect()
}

fn render_document(value: &Value) -> Vec<String> {
    if let Some(items) = value.as_array() {
        return items.iter().filter_map(render_entry).collect();
    }
    for key in ["messages", "history", "chat", "turns", "items"] {
        if let Some(items) = value.get(key).and_then(Value::as_array) {
            return items.iter().filter_map(render_entry).collect();
        }
    }
    Vec::new()
}

/// One transcript entry → one prose line, or nothing.
pub fn render_entry(value: &Value) -> Option<String> {
    // Codex wraps everything in `{type, payload}`; unwrap once.
    if let Some(payload) = value.get("payload").filter(|payload| payload.is_object()) {
        return render_entry(payload);
    }
    let role = value
        .get("type")
        .and_then(Value::as_str)
        .filter(|kind| matches!(*kind, "user" | "assistant" | "system"))
        .or_else(|| value.get("role").and_then(Value::as_str))?;

    let content = value
        .get("message")
        .and_then(|message| message.get("content"))
        .or_else(|| value.get("content"))
        .or_else(|| value.get("text"))?;

    let body = render_content(content);
    let body = body.trim();
    if body.is_empty() {
        return None;
    }
    let label = match role {
        "user" => "[用户]",
        "assistant" => "[助手]",
        _ => "[系统]",
    };
    // A tool line already carries its own label.
    if body.starts_with("[工具") || body.starts_with("[结果") {
        return Some(clamp(body));
    }
    Some(clamp(&format!("{label} {body}")))
}

/// `content` is a string in some CLIs and a block array in others.
fn render_content(content: &Value) -> String {
    match content {
        Value::String(text) => collapse(text),
        Value::Array(blocks) => blocks
            .iter()
            .filter_map(render_block)
            .collect::<Vec<_>>()
            .join(" "),
        Value::Object(_) => render_block(content).unwrap_or_default(),
        _ => String::new(),
    }
}

fn render_block(block: &Value) -> Option<String> {
    let kind = block.get("type").and_then(Value::as_str).unwrap_or("text");
    match kind {
        "text" | "output_text" | "input_text" => block
            .get("text")
            .and_then(Value::as_str)
            .map(collapse)
            .filter(|text| !text.is_empty()),
        "tool_use" | "function_call" => {
            let name = block
                .get("name")
                .and_then(Value::as_str)
                .unwrap_or("未命名工具");
            let detail = block
                .get("input")
                .or_else(|| block.get("arguments"))
                .map(summarize_input)
                .unwrap_or_default();
            Some(if detail.is_empty() {
                format!("[工具 {name}]")
            } else {
                format!("[工具 {name} {detail}]")
            })
        }
        "tool_result" | "function_call_output" => {
            let detail = block.get("content").map(render_content).unwrap_or_default();
            Some(format!(
                "[结果 {}]",
                shorten(detail.trim(), MAX_TOOL_DETAIL)
            ))
        }
        "thinking" | "redacted_thinking" => None,
        _ => None,
    }
}

/// A tool's arguments reduced to the one field a reader cares about.
fn summarize_input(input: &Value) -> String {
    for key in [
        "file_path",
        "path",
        "command",
        "pattern",
        "query",
        "url",
        "description",
        "prompt",
    ] {
        if let Some(value) = input.get(key).and_then(Value::as_str) {
            return shorten(&collapse(value), MAX_TOOL_DETAIL);
        }
    }
    match input {
        Value::String(text) => shorten(&collapse(text), MAX_TOOL_DETAIL),
        Value::Object(map) if map.is_empty() => String::new(),
        other => shorten(&collapse(&other.to_string()), MAX_TOOL_DETAIL),
    }
}

fn collapse(text: &str) -> String {
    text.split_whitespace().collect::<Vec<_>>().join(" ")
}

fn shorten(text: &str, max_chars: usize) -> String {
    if text.chars().count() <= max_chars {
        return text.to_owned();
    }
    let head: String = text.chars().take(max_chars).collect();
    format!("{head}…")
}

fn clamp(line: &str) -> String {
    shorten(line, MAX_LINE)
}

/* -------------------------------- locating -------------------------------- */

/// Finds the transcript for a node. `transcript_path` is what the CLI itself
/// reported (claude); the others are found by session id under their own
/// config home.
pub fn locate(
    agent_id: &str,
    transcript_path: Option<&str>,
    session_id: Option<&str>,
) -> Option<Located> {
    if let Some(path) = transcript_path
        .map(PathBuf::from)
        .filter(|path| path.is_file())
    {
        let origin = format!("转录文件 {}", path.display());
        return Some(Located { path, origin });
    }
    let session_id = session_id?;
    match agent_id {
        "codex" => find_under(&codex_home().join("sessions"), |name| {
            name.starts_with("rollout-") && name.contains(session_id) && name.ends_with(".jsonl")
        })
        .map(|path| Located {
            origin: format!("Codex 会话记录 {}", path.display()),
            path,
        }),
        "gemini" => find_under(&gemini_home().join("tmp"), |name| {
            name.contains(session_id) && name.ends_with(".json")
        })
        .map(|path| Located {
            origin: format!("Gemini 会话记录 {}", path.display()),
            path,
        }),
        _ => None,
    }
}

fn home() -> PathBuf {
    std::env::var_os("HOME")
        .or_else(|| std::env::var_os("USERPROFILE"))
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("."))
}

pub fn codex_home() -> PathBuf {
    std::env::var_os("CODEX_HOME")
        .map(PathBuf::from)
        .filter(|path| !path.as_os_str().is_empty())
        .unwrap_or_else(|| home().join(".codex"))
}

/// `~/.gemini`, or `$GEMINI_CLI_HOME/.gemini`.
///
/// `GEMINI_CLI_HOME` overrides the *home directory* the CLI reads from, not the
/// `.gemini` folder inside it: gemini-cli's own `homedir()` returns the
/// variable and `Storage::getGlobalGeminiDir()` joins `.gemini` onto the
/// result. Reading the variable as the config directory itself sent every
/// transcript lookup one level too high, where nothing matches and the honest
/// answer "no transcript" is indistinguishable from a real absence.
pub fn gemini_home() -> PathBuf {
    gemini_home_in(
        std::env::var_os("GEMINI_CLI_HOME")
            .map(PathBuf::from)
            .filter(|path| !path.as_os_str().is_empty()),
        &home(),
    )
}

/// The rule on its own, with the environment passed in so a test can state it
/// without mutating a process-global the rest of the suite also reads.
pub fn gemini_home_in(cli_home: Option<PathBuf>, home: &Path) -> PathBuf {
    cli_home.unwrap_or_else(|| home.to_path_buf()).join(".gemini")
}

/// A bounded breadth-first walk. Neither CLI documents its directory layout, so
/// the search is by file name; the bounds are what keep a surprising layout
/// (a symlink loop, a million-file cache) from turning a read into a hang.
pub fn find_under(root: &Path, matches: impl Fn(&str) -> bool) -> Option<PathBuf> {
    if !root.is_dir() {
        return None;
    }
    let mut frontier = vec![(root.to_path_buf(), 0usize)];
    let mut seen = 0usize;
    let mut best: Option<(std::time::SystemTime, PathBuf)> = None;
    while let Some((directory, depth)) = frontier.pop() {
        let Ok(entries) = std::fs::read_dir(&directory) else {
            continue;
        };
        for entry in entries.flatten() {
            seen += 1;
            if seen > MAX_SCAN_ENTRIES {
                return best.map(|(_, path)| path);
            }
            let path = entry.path();
            let Ok(kind) = entry.file_type() else {
                continue;
            };
            if kind.is_dir() {
                if depth < MAX_SCAN_DEPTH {
                    frontier.push((path, depth + 1));
                }
                continue;
            }
            let name = entry.file_name();
            let Some(name) = name.to_str() else { continue };
            if !matches(name) {
                continue;
            }
            let modified = entry
                .metadata()
                .and_then(|metadata| metadata.modified())
                .unwrap_or(std::time::SystemTime::UNIX_EPOCH);
            if best.as_ref().is_none_or(|(best, _)| modified > *best) {
                best = Some((modified, path));
            }
        }
    }
    best.map(|(_, path)| path)
}

/// opencode keeps its history in its own store; the CLI is the only reader.
pub async fn opencode_export(session_id: &str) -> Option<String> {
    let binary = crate::agent::resolve_command("opencode")?;
    let output = tokio::time::timeout(
        std::time::Duration::from_secs(10),
        tokio::process::Command::new(binary)
            .arg("export")
            .arg(session_id)
            .kill_on_drop(true)
            .output(),
    )
    .await
    .ok()?
    .ok()?;
    if !output.status.success() {
        return None;
    }
    Some(String::from_utf8_lossy(&output.stdout).into_owned())
}

File: apps/desktop/src/core/collab/transcript.ts
import { closeSync, openSync, readSync, readdirSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

/**
 * Locating and rendering another agent's transcript.
 *
 * Ported from the pre-merge implementation. Every CLI writes its
 * history somewhere different and none of them promise the shape, so this
 * module is deliberately forgiving: it reads the tail of a file, renders the
 * lines it recognises, and skips the ones it does not. A transcript we can
 * only half read is far more useful to the agent asking than an error, and one
 * we cannot find at all is reported in a sentence rather than as a failure.
 */

/** Only the tail of a transcript is read; a long session is megabytes of JSON. */
export const MAX_TAIL_BYTES = 5 * 1024 * 1024;
/** `transcript` renders everything it found, up to this many bytes of prose. */
export const MAX_RENDERED_BYTES = 200 * 1024;
/** A tool's input is quoted, not dumped. */
const MAX_TOOL_DETAIL = 120;
/** One rendered message line is trimmed to this before it reaches the agent. */
const MAX_LINE = 2_000;
/** Ceiling on how many directory entries a session-id search will look at. */
const MAX_SCAN_ENTRIES = 20_000;
const MAX_SCAN_DEPTH = 6;

/** Where a transcript came from, so the reply can say so. */
export interface Located {
  readonly path: string;
  /** Human sentence naming the provider and file. */
  readonly origin: string;
}

/**
 * Reads at most the last `maxBytes` of a file, starting at the first newline
 * inside the window so the first line is never a fragment.
 */
export function readTail(path: string, maxBytes: number): string {
  const handle = openSync(path, "r");
  try {
    const length = statSync(path).size;
    const start = Math.max(0, length - maxBytes);
    const size = Math.min(length - start, maxBytes);
    const buffer = Buffer.alloc(size);
    let filled = 0;
    while (filled < size) {
      const read = readSync(
        handle,
        buffer,
        filled,
        size - filled,
        start + filled,
      );
      if (read === 0) break;
      filled += read;
    }
    const text = buffer.subarray(0, filled).toString("utf8");
    if (start === 0) return text;
    const newline = text.indexOf("\n");
    return newline === -1 ? "" : text.slice(newline + 1);
  } finally {
    closeSync(handle);
  }
}

/**
 * 渲染的两档（设计 `agent-delivery.md` §13 第 2 条）。
 *
 * 缺省这一档就是这个模块一直以来的行为，所以 `render(text)` 一个字都不用改。
 * `transcript` 那条路传的是收紧过的一档：每条消息截到 2 KB，`tool_result` 只
 * 留工具名、字节数与首行——工具结果是一份转录里最长、对读者最没用的一段，一
 * 次 `Read` 的回显就能顶掉整份预算。
 */
export interface RenderOptions {
  /** 一条消息截到这么多个字符。 */
  readonly maxLineChars?: number;
  /** `tool_result` 只留工具名、字节数与首行。 */
  readonly briefToolResults?: boolean;
}

interface RenderContext {
  readonly maxLineChars: number;
  readonly briefToolResults: boolean;
  /**
   * `tool_use_id` → 工具名，跨条目累积。
   *
   * `tool_result` 自己不带工具名，只带它回应的那次调用的 id；要说出「这是哪个
   * 工具的结果」就得记得前面那条 `tool_use`。一份转录是按时间顺序读的，所以这
   * 张表只需要往前看。
   */
  readonly toolNames: Map<string, string>;
}

function contextOf(options: RenderOptions | undefined): RenderContext {
  return {
    maxLineChars: options?.maxLineChars ?? MAX_LINE,
    briefToolResults: options?.briefToolResults ?? false,
    toolNames: new Map(),
  };
}

/**
 * 一条渲染好的记录，外加它在源文本里结束于第几个字节。
 *
 * 那个偏移是增量游标的全部实现（§13 第 3 条）：交出去 N 条之后，下次从第 N 条
 * 之后那个字节接着读。它算的是**源文本**的字节，不是渲染出来的散文的字节。
 */
export interface TranscriptRecord {
  readonly line: string;
  readonly endOffset: number;
}

/** 一次增量读取的结果：读到的文本，以及现在的文件尾在第几个字节。 */
export interface Range {
  readonly text: string;
  readonly startOffset: number;
  readonly endOffset: number;
}

/**
 * 从 `startByte` 读到文件尾，最多 `maxBytes`（增量游标那条路）。
 *
 * 偏移大于文件长度的时候从头读：转录被换掉或者被截短了，那个偏移在新内容里指
 * 的是另一段话。路径是否还是同一个由游标自己判（`context-reads.ts`），长度这
 * 一层的判据在这里。
 */
export function readRange(
  path: string,
  startByte: number,
  maxBytes: number,
): Range {
  const handle = openSync(path, "r");
  try {
    const length = statSync(path).size;
    const start = startByte > length || startByte < 0 ? 0 : startByte;
    // 超过上限时保留**尾部**：新的那些比旧的那些有用。
    const from = Math.max(start, length - maxBytes);
    const size = Math.max(0, length - from);
    if (size === 0) return { text: "", startOffset: from, endOffset: length };
    const buffer = Buffer.alloc(size);
    let filled = 0;
    while (filled < size) {
      const read = readSync(
        handle,
        buffer,
        filled,
        size - filled,
        from + filled,
      );
      if (read === 0) break;
      filled += read;
    }
    return {
      text: buffer.subarray(0, filled).toString("utf8"),
      startOffset: from,
      endOffset: length,
    };
  } finally {
    closeSync(handle);
  }
}

/**
 * Renders JSONL (or a JSON array / object of messages) into one line per
 * message. Unknown lines are skipped rather than reported.
 */
export function render(text: string, options?: RenderOptions): string[] {
  return renderRecords(text, options).map((record) => record.line);
}

/** {@link render}，但每条还带着它在源文本里的结束偏移。 */
export function renderRecords(
  text: string,
  options?: RenderOptions,
): TranscriptRecord[] {
  const context = contextOf(options);
  const total = Buffer.byteLength(text, "utf8");
  const trimmed = text.trimStart();
  if (trimmed.startsWith("[") || trimmed.startsWith("{")) {
    let value: unknown;
    try {
      value = JSON.parse(trimmed);
    } catch {
      value = undefined;
    }
    if (value !== undefined) {
      const rendered = renderDocument(value, context);
      // 整份 JSON 没有「读到一半」这回事：偏移一律是文件尾。
      if (rendered.length > 0) {
        return rendered.map((line) => ({ line, endOffset: total }));
      }
    }
  }
  const records: TranscriptRecord[] = [];
  let offset = 0;
  for (const raw of text.split("\n")) {
    // `+1` 是被 `split` 吃掉的那个换行；最后一段多算一个字节不影响判据（游标
    // 只会因此少读零字节），但少算会让同一条被读第二次。
    offset += Buffer.byteLength(raw, "utf8") + 1;
    const line = raw.trim();
    if (line === "") continue;
    let value: unknown;
    try {
      value = JSON.parse(line);
    } catch {
      continue;
    }
    const rendered = renderEntry(value, context);
    if (rendered !== undefined) {
      records.push({ line: rendered, endOffset: Math.min(offset, total) });
    }
  }
  return records;
}

function renderDocument(value: unknown, context: RenderContext): string[] {
  const render = (entry: unknown): string | undefined =>
    renderEntry(entry, context);
  if (Array.isArray(value)) {
    return value
      .map(render)
      .filter((line): line is string => line !== undefined);
  }
  if (value === null || typeof value !== "object") return [];
  const record = value as Record<string, unknown>;
  for (const key of ["messages", "history", "chat", "turns", "items"]) {
    const items = record[key];
    if (Array.isArray(items)) {
      return items
        .map(render)
        .filter((line): line is string => line !== undefined);
    }
  }
  return [];
}

const ROLES = ["user", "assistant", "system"];

/** One transcript entry → one prose line, or nothing. */
export function renderEntry(
  value: unknown,
  options?: RenderOptions | RenderContext,
): string | undefined {
  const context = isContext(options) ? options : contextOf(options);
  if (value === null || typeof value !== "object") return undefined;
  const record = value as Record<string, unknown>;
  // Codex wraps everything in `{type, payload}`; unwrap once.
  const payload = record.payload;
  if (
    payload !== null &&
    typeof payload === "object" &&
    !Array.isArray(payload)
  ) {
    return renderEntry(payload, context);
  }
  const kind = record.type;
  const role =
    typeof kind === "string" && ROLES.includes(kind)
      ? kind
      : typeof record.role === "string"
        ? record.role
        : undefined;
  if (role === undefined) return undefined;

  const message = record.message;
  const content =
    message !== null && typeof message === "object" && !Array.isArray(message)
      ? ((message as Record<string, unknown>).content ??
        record.content ??
        record.text)
      : (record.content ?? record.text);
  if (content === undefined || content === null) return undefined;

  const body = renderContent(content, context).trim();
  if (body === "") return undefined;
  const label =
    role === "user" ? "[用户]" : role === "assistant" ? "[助手]" : "[系统]";
  // A tool line already carries its own label.
  if (body.startsWith("[工具") || body.startsWith("[结果")) {
    return clamp(body, context.maxLineChars);
  }
  return clamp(`${label} ${body}`, context.maxLineChars);
}

function isContext(
  options: RenderOptions | RenderContext | undefined,
): options is RenderContext {
  return options !== undefined && "toolNames" in options;
}

/** `content` is a string in some CLIs and a block array in others. */
function renderContent(content: unknown, context: RenderContext): string {
  if (typeof content === "string") return collapse(content);
  if (Array.isArray(content)) {
    return content
      .map((block) => renderBlock(block, context))
      .filter((part): part is string => part !== undefined)
      .join(" ");
  }
  if (content !== null && typeof content === "object") {
    return renderBlock(content, context) ?? "";
  }
  return "";
}

function renderBlock(
  block: unknown,
  context: RenderContext,
): string | undefined {
  if (block === null || typeof block !== "object") return undefined;
  const record = block as Record<string, unknown>;
  const kind = typeof record.type === "string" ? record.type : "text";
  switch (kind) {
    case "text":
    case "output_text":
    case "input_text": {
      const text = record.text;
      if (typeof text !== "string") return undefined;
      const collapsed = collapse(text);
      return collapsed === "" ? undefined : collapsed;
    }
    case "tool_use":
    case "function_call": {
      const name = typeof record.name === "string" ? record.name : "未命名工具";
      const id = record.id ?? record.tool_use_id ?? record.call_id;
      if (typeof id === "string" && id !== "") context.toolNames.set(id, name);
      const input = record.input ?? record.arguments;
      const detail = input === undefined ? "" : summarizeInput(input);
      return detail === "" ? `[工具 ${name}]` : `[工具 ${name} ${detail}]`;
    }
    case "tool_result":
    case "function_call_output": {
      const detail =
        record.content === undefined
          ? ""
          : typeof record.content === "string"
            ? record.content
            : renderContent(record.content, context);
      if (!context.briefToolResults) {
        return `[结果 ${shorten(collapse(detail), MAX_TOOL_DETAIL)}]`;
      }
      const id = record.tool_use_id ?? record.call_id ?? record.id;
      const name =
        (typeof id === "string" ? context.toolNames.get(id) : undefined) ??
        "工具";
      const bytes = Buffer.byteLength(detail, "utf8");
      const first = collapse(detail.split("\n")[0] ?? "");
      return `[结果 ${name} ${bytes} B${first === "" ? "" : ` 首行：${shorten(first, MAX_TOOL_DETAIL)}`}]`;
    }
    // Thinking blocks are the model talking to itself; not somebody else's
    // context.
    default:
      return undefined;
  }
}

/** A tool's arguments reduced to the one field a reader cares about. */
function summarizeInput(input: unknown): string {
  if (input !== null && typeof input === "object" && !Array.isArray(input)) {
    const record = input as Record<string, unknown>;
    for (const key of [
      "file_path",
      "path",
      "command",
      "pattern",
      "query",
      "url",
      "description",
      "prompt",
    ]) {
      const value = record[key];
      if (typeof value === "string") {
        return shorten(collapse(value), MAX_TOOL_DETAIL);
      }
    }
    if (Object.keys(record).length === 0) return "";
  }
  if (typeof input === "string")
    return shorten(collapse(input), MAX_TOOL_DETAIL);
  return shorten(collapse(JSON.stringify(input) ?? ""), MAX_TOOL_DETAIL);
}

function collapse(text: string): string {
  return text.trim().split(/\s+/).filter(Boolean).join(" ");
}

function shorten(text: string, maxChars: number): string {
  const characters = [...text];
  if (characters.length <= maxChars) return text;
  return `${characters.slice(0, maxChars).join("")}…`;
}

function clamp(line: string, maxChars: number): string {
  return shorten(line, maxChars);
}

/* -------------------------------- locating -------------------------------- */

export function home(): string {
  return process.env.HOME ?? process.env.USERPROFILE ?? homedir() ?? ".";
}

export function codexHome(): string {
  const configured = process.env.CODEX_HOME;
  if (configured !== undefined && configured !== "") return configured;
  return join(home(), ".codex");
}

/**
 * Finds the transcript for a node.
 *
 * `transcriptPath` is what the CLI itself reported (claude); the others are
 * found by session id under their own config home. A provider with neither is
 * `undefined`, which the caller must report as "this CLI keeps nothing
 * readable" rather than as an empty conversation.
 */
export function locate(
  agentId: string,
  transcriptPath: string | undefined,
  sessionId: string | undefined,
): Located | undefined {
  if (transcriptPath !== undefined && isFile(transcriptPath)) {
    return { path: transcriptPath, origin: `转录文件 ${transcriptPath}` };
  }
  if (sessionId === undefined || sessionId === "") return undefined;
  if (agentId !== "codex") return undefined;
  const found = findUnder(
    join(codexHome(), "sessions"),
    (name) =>
      name.startsWith("rollout-") &&
      name.includes(sessionId) &&
      name.endsWith(".jsonl"),
  );
  return found === undefined
    ? undefined
    : { path: found, origin: `Codex 会话记录 ${found}` };
}

function isFile(path: string): boolean {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

/**
 * A bounded walk that answers with the newest matching file.
 *
 * Codex does not document its directory layout, so the search is by file name;
 * the bounds are what keep a surprising layout (a symlink loop, a
 * million-file cache) from turning a read into a hang.
 */
export function findUnder(
  root: string,
  matches: (name: string) => boolean,
): string | undefined {
  if (!isDirectory(root)) return undefined;
  const frontier: [string, number][] = [[root, 0]];
  let seen = 0;
  let best: { modified: number; path: string } | undefined;
  while (frontier.length > 0) {
    const [directory, depth] = frontier.pop() as [string, number];
    let entries;
    try {
      entries = readdirSync(directory, { withFileTypes: true });
    } catch {
      continue;
    }
    for (const entry of entries) {
      seen += 1;
      if (seen > MAX_SCAN_ENTRIES) return best?.path;
      const path = join(directory, entry.name);
      if (entry.isDirectory()) {
        if (depth < MAX_SCAN_DEPTH) frontier.push([path, depth + 1]);
        continue;
      }
      if (!entry.isFile() || !matches(entry.name)) continue;
      let modified = 0;
      try {
        modified = statSync(path).mtimeMs;
      } catch {
        modified = 0;
      }
      if (best === undefined || modified > best.modified) {
        best = { modified, path };
      }
    }
  }
  return best?.path;
}

function isDirectory(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}
